[Generic Signature Builder] Improve the handling of typealiases in protocols.

huonw opened the pull request:
Previously, the following example

    protocol P {
        associatedtype A
        typealias X = A
    }
    protocol Q {
        associatedtype B: P
    }

    func requirementOnNestedTypeAlias<T>(_: T) where T: Q, T.B.X == Int {}

would result in the T.B.X == Int requirement being dropped, because it adopted the (Protocol) RequirementSource of the T.B.X == T.B.A same-type requirement implied by the typealias declaration.

DougGregor commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -587,13 +585,21 @@ auto ArchetypeBuilder::PotentialArchetype::getNestedType(
+        SubstitutionMap subMap;
+        subMap.addSubstitution(self, concreteSelf);
+        subMap.addConformance(self, ProtocolConformanceRef(proto));
+        auto typeHere = typeInProtocol.subst(subMap);
```
@slavapestov is going to be grouchy at building the substitution map this way. Can you just provide the trivial lookup functions that handle "self -> concreteSelf" and the construction of the abstract conformance? There's no need to build a full substitution map here.

huonw replied:
Oh, I didn't realize I could subst with a closure.

slavapestov replied:
Even better would be to use the new SubstitutionMap::getProtocolSubstitutions() function. It handles exactly this case where you're replacing Self and you have a conformance.

huonw replied:
Thanks! I noticed that and have it in my local branch already... currently trying to make the rest of the patch work properly/clean.

DougGregor commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -1030,6 +1036,37 @@ auto ArchetypeBuilder::resolveArchetype(Type type, PotentialArchetype *basePA)
+  // (e.g. protocol P { associatedtype A; typealias B = A }), so we need to
+  // resolve that.
+  auto concrete = pa->getConcreteType();
+  auto rhsPA = resolveArchetype(concrete, basePA);
```
The assertion above is the reason we don't need to recursively call `resolveArchetypeOrTypeThroughTypeAlias` here, right?

huonw replied:
Yes; I'll add a comment.

DougGregor commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -1030,6 +1036,37 @@ auto ArchetypeBuilder::resolveArchetype(Type type, PotentialArchetype *basePA)
+    return type;
+
+  pa = pa->getRepresentative();
+  if (!pa->getParent() || !pa->getTypeAliasDecl())
```
Hmm... it occurs to me that there might be direct callers to `addSameTypeRequirementBetweenArchetypes` that unknowingly end up with a `typealias` for one of the archetypes. Should this logic be sunk down into that function instead?

huonw replied:
I think the only remaining direct callers are either `addSameTypeRequirement`, or are in places where equating to the type alias PA directly seems correct, specifically: potential archetype typo correction, equating any new PAs inside `getNestedType` itself, and recursively to merge the PAs.

However, having it inside `addSameTypeRequirementBetweenArchetypes` does seem like it might be nicer.

huonw committed:
[Generic signature builder] resolve types deeper in same type requirements.

This adds the concept of an "unresolved type" and a "resolved type",
where the latter involves stripping away any typealias archetype layers
from the former, to get to the "true" potential archetype or concrete
type that something refers to.

huonw commented:
I've updated this in a fairly invasive way.

DougGregor reviewed (approved):
This is way cleaner. Looks great!

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Looks like some compiler crashers regressed due to the `llvm_unreachable`.

huonw commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -177,6 +177,40 @@ void GenericSignatureBuilder::PotentialArchetype::resolveAssociatedType(
+  PotentialArchetype *getPotentialArchetype() const {
+    return paOrT.dyn_cast<PotentialArchetype *>();
+  }
+  bool isUnresolvedTypeAlias() const { return unresolvedTypeAlias; }
```
Note to self: this needs renaming.

huonw commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -1001,6 +1031,39 @@ auto GenericSignatureBuilder::resolveArchetype(Type type) -> PotentialArchetype
+  // (e.g. protocol P { associatedtype A; typealias B = A }), so we need to
+  // resolve that.  However, the archetype should always be resolved far enough
+  // upon creation to not be another type alias (verified by the assertion
+  // below), and hence this function doesn't need to be recursive..
```
Note to self: this comment is outdated.

huonw committed:
[Generic signature builder] diagnose typealiases with the same name and different types.

At the moment, these are only done on demand, when the typealias is
actually used somewhere, because there's currently a recursive
validation loop that stops doing it more eagerly from working in many
cases.

huonw committed:
[Generic signature builder] paper over problems with generic typealiases in protocols.

Generic typealiases are entirely broken as archetypes, but some things
mostly worked with them before introducing ResolvedType, so let's
restore that behaviour.

huonw commented:
Another update, removing the reimplementation of #7563, among other things.

There's an unfortunate FIXME in `getNestedType` that is stopping diagnostics about typealiases with mismatching types from being emitted eagerly, instead they are currently only emitted when the typealias is used.

huonw commented:
@swift-ci Please test

DougGregor commented:
I recommend smoke tests :)

huonw commented:
@swift-ci Please smoke test and merge

DougGregor commented:
@swift-ci please smoke test Linux
